src/Abstract/Type.hs

```diff
@@ -1,10 +1,47 @@
+{-# LANGUAGE FlexibleContexts, FlexibleInstances, GADTs, MultiParamTypeClasses, TypeOperators, UndecidableInstances #-}
 module Abstract.Type where
 
+import Abstract.Util
+import Control.Effect
+import Control.Monad.Effect.Internal
+import Control.Monad.Fail
 import Data.Text.Prettyprint.Doc
+import Prelude hiding (fail)
 
-data Type = Int | Bool | Type :-> Type | Type :* Type
+type TName = Int
+
+data Type = Int | Bool | Type :-> Type | Type :* Type | TVar TName
   deriving (Eq, Ord, Show)
 
 
+unify :: MonadFail m => Type -> Type -> m Type
+unify Int  Int  = pure Int
+unify Bool Bool = pure Bool
+unify (a1 :-> b1) (a2 :-> b2) = (:->) <$> unify a1 a2 <*> unify b1 b2
+unify (a1 :* b1)  (a2 :* b2)  = (:*)  <$> unify a1 a2 <*> unify b1 b2
+unify (TVar _) b = pure b
+unify a (TVar _) = pure a
+unify t1 t2 = fail ("cannot unify " ++ prettyString t1 ++ " with " ++ prettyString t2)
```

Comment:
This is probably a hint that we should be using `MonadError` instead of `MonadFail`.